storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -95,13 +108,86 @@ func (s *commitLogSource) Read(
 	defer iter.Close()
 
 	var (
-		unmerged    = make(map[uint32]map[ts.Hash]encoderMap)
+		numShards = len(shardsTimeRanges)
+		unmerged  = make([]map[ts.Hash]encoderMap, numShards, numShards)
+		// TODO: Add to opts
+		numConc     = uint32(4)
 		bopts       = s.opts.ResultOptions()
 		blopts      = bopts.DatabaseBlockOptions()
 		blockSize   = bopts.RetentionOptions().BlockSize()
 		encoderPool = bopts.DatabaseBlockOptions().EncoderPool()
-		errs        = 0
 	)
+
+	encoderChans := []chan encoderArg{}
+	for i := uint32(0); i < numConc; i++ {
+		encoderChans = append(encoderChans, make(chan encoderArg, 1000))
+	}
+
+	// Spin up numConc background go-routines to handle M3TSZ encoding. This must
+	// happen before we start reading to prevent infinitely blocking writes to
+	// the encoderChans.
+	wg := sync.WaitGroup{}
+	workerErrs := make([]int, numConc, numConc)
```

Comment:
Not related to your change, but considering you're reworking this part, it'd be good to refactor this method into smaller chunks.